File: src/modules/sage_infall/test_sage_infall.py
#!/usr/bin/env python3
"""
Integration tests for sage_infall module.

This test validates software quality aspects of the sage_infall module:
- Module loads and initializes correctly
- Parameters can be configured via .par files
- Module executes without errors or memory leaks
- Output properties appear in output files
- Module works in multi-module pipelines

NOTE: Physics validation (reionization correctness, infall amounts) deferred
      to Phase 4.3+ when downstream modules (cooling, star formation) are implemented.

Phase: Phase 4.2 (SAGE Physics Module Implementation)
Author: Mimic Development Team
Date: 2025-11-12
"""

import os
import sys
import shutil
import subprocess
import tempfile
import unittest

# Add tests directory to path to import framework
repo_root = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../..'))
sys.path.insert(0, os.path.join(repo_root, 'tests'))
from framework import load_binary_halos


class TestSageInfall(unittest.TestCase):
    """Integration tests for sage_infall module software quality."""

    @classmethod
    def setUpClass(cls):
        """Set up test environment once for all tests."""
        # Get repository root (three levels up from src/modules/sage_infall/)
        cls.repo_root = os.path.abspath(
            os.path.join(os.path.dirname(__file__), "..", "..", "..")
        )

        # Path to mimic executable
        cls.mimic_exe = os.path.join(cls.repo_root, "mimic")

        # Check mimic is compiled
        if not os.path.exists(cls.mimic_exe):
            raise RuntimeError(
                f"Mimic executable not found at {cls.mimic_exe}. "
                "Run 'make' to compile."
            )

        # Reference parameter file
        cls.ref_param_file = os.path.join(
            cls.repo_root, "input", "millennium.par"
        )

        # Create temporary directory for test outputs
        cls.temp_dir = tempfile.mkdtemp(prefix="mimic_sage_infall_test_")

    @classmethod
    def tearDownClass(cls):
        """Clean up test environment."""
        # Remove temporary directory
        if os.path.exists(cls.temp_dir):
            shutil.rmtree(cls.temp_dir)

    def _read_param_file(self, param_file):
        """Read parameter file and return as dictionary."""
        params = {}
        with open(param_file, 'r') as f:
            for line in f:
                line = line.strip()
                # Skip comments and empty lines
                if not line or line.startswith('%') or line.startswith('#'):
                    continue
                # Handle arrow notation (-> means skip)
                if '->' in line:
                    continue
                # Parse key-value pairs
                parts = line.split()
                if len(parts) >= 2:
                    params[parts[0]] = ' '.join(parts[1:])
        return params

    def _create_test_param_file(self, output_name, enabled_modules=None,
                                module_params=None, first_file=0, last_file=0):
        """
        Create a test parameter file with specified module configuration.

        Args:
            output_name: Name for output directory
            enabled_modules: List of module names to enable
            module_params: Dict of {ParamName: value} for module parameters
            first_file: First file to process (default: 0)
            last_file: Last file to process (default: 0)

        Returns:
            Path to created parameter file
        """
        # Read reference parameter file
        ref_params = self._read_param_file(self.ref_param_file)

        # Create output directory
        output_dir = os.path.join(self.temp_dir, output_name)
        os.makedirs(output_dir, exist_ok=True)

        # Update parameters
        ref_params['OutputDir'] = output_dir
        ref_params['FirstFile'] = str(first_file)
        ref_params['LastFile'] = str(last_file)
        ref_params['OutputFormat'] = 'binary'  # Use binary format for testing

        # Write test parameter file
        param_path = os.path.join(self.temp_dir, f"{output_name}.par")
        with open(param_path, 'w') as f:
            f.write("%------------------------------------------\n")
            f.write("%----- sage_infall Integration Test ------\n")
            f.write("%------------------------------------------\n\n")

            # Write basic parameters
            for key in ['OutputFileBaseName', 'OutputDir', 'FirstFile',
                       'LastFile', 'TreeName', 'TreeType', 'OutputFormat',
                       'SimulationDir', 'FileWithSnapList', 'LastSnapshotNr',
                       'BoxSize', 'Omega', 'OmegaLambda', 'Hubble_h',
                       'PartMass', 'UnitLength_in_cm', 'UnitMass_in_g',
                       'UnitVelocity_in_cm_per_s']:
                if key in ref_params:
                    f.write(f"{key:30s} {ref_params[key]}\n")

            # Write snapshot output list
            f.write("\nNumOutputs        8\n")
            f.write("-> 63 37 32 27 23 20 18 16\n\n")

            # Write module configuration
            f.write("%------------------------------------------\n")
            f.write("%----- Galaxy Physics Modules -------------\n")
            f.write("%------------------------------------------\n")

            if enabled_modules:
                f.write(f"EnabledModules          {','.join(enabled_modules)}\n\n")

                # Write module parameters
                if module_params:
                    for param_name, value in module_params.items():
                        f.write(f"{param_name:30s} {value}\n")

        return param_path

    def _run_mimic(self, param_file):
        """
        Run mimic with specified parameter file.

        Returns:
            (returncode, stdout, stderr)
        """
        result = subprocess.run(
            [self.mimic_exe, param_file],
            capture_output=True,
            text=True,
            timeout=60  # 60 second timeout for integration tests
        )
        return result.returncode, result.stdout, result.stderr

    def test_module_loads(self):
        """Test that sage_infall module loads and initializes successfully."""
        print("\n  Testing module load and initialization...")

        param_file = self._create_test_param_file(
            output_name="sage_infall_load",
            enabled_modules=["sage_infall"],
            module_params={
                "SageInfall_BaryonFrac": "0.17",
                "SageInfall_ReionizationOn": "1"
            }
        )

        returncode, stdout, stderr = self._run_mimic(param_file)

        # Check execution succeeded
        self.assertEqual(returncode, 0,
                        f"Mimic should execute successfully with sage_infall\nStderr: {stderr}")

        # Check initialization log message
        self.assertIn("SAGE infall module initialized", stdout,
                     "sage_infall should log initialization message")

    def test_output_properties_exist(self):
        """Test that HotGas and related properties appear in output."""
        print("\n  Testing output properties...")

        param_file = self._create_test_param_file(
            output_name="sage_infall_output",
            enabled_modules=["sage_infall"]
        )

        returncode, stdout, stderr = self._run_mimic(param_file)
        self.assertEqual(returncode, 0, "Mimic execution should succeed")

        # Load output file
        output_dir = os.path.join(self.temp_dir, "sage_infall_output")
        output_file = os.path.join(output_dir, "model_z0.000_0")
        self.assertTrue(os.path.exists(output_file), "Output file should exist")

        # Load and check halos
        halos, metadata = load_binary_halos(output_file)
        self.assertGreater(len(halos), 0, "Should have halos in output")

        # Check HotGas property exists
        self.assertIn('HotGas', halos.dtype.names,
                     "HotGas property should exist in output")
        self.assertIn('MetalsHotGas', halos.dtype.names,
                     "MetalsHotGas property should exist in output")

    def test_parameters_configurable(self):
        """Test that all sage_infall parameters can be configured."""
        print("\n  Testing parameter configuration...")

        # Test with non-default parameter values
        param_file = self._create_test_param_file(
            output_name="sage_infall_params",
            enabled_modules=["sage_infall"],
            module_params={
                "SageInfall_BaryonFrac": "0.20",
                "SageInfall_ReionizationOn": "0",
                "SageInfall_Reionization_z0": "9.0",
                "SageInfall_Reionization_zr": "6.0",
                "SageInfall_StrippingSteps": "5"
            }
        )

        returncode, stdout, stderr = self._run_mimic(param_file)
        self.assertEqual(returncode, 0, "Execution with custom parameters should succeed")

        # Verify parameters were read
        self.assertIn("BaryonFrac = 0.2000", stdout,
                     "Custom BaryonFrac should be logged")
        self.assertIn("ReionizationOn = 0", stdout,
                     "Custom ReionizationOn should be logged")

    def test_reionization_toggle(self):
        """Test that ReionizationOn parameter affects execution (no crash)."""
        print("\n  Testing reionization toggle...")

        # Test with reionization ON
        param_file_on = self._create_test_param_file(
            output_name="sage_infall_reion_on",
            enabled_modules=["sage_infall"],
            module_params={"SageInfall_ReionizationOn": "1"}
        )

        returncode_on, stdout_on, stderr_on = self._run_mimic(param_file_on)
        self.assertEqual(returncode_on, 0, "Should run with reionization ON")

        # Test with reionization OFF
        param_file_off = self._create_test_param_file(
            output_name="sage_infall_reion_off",
            enabled_modules=["sage_infall"],
            module_params={"SageInfall_ReionizationOn": "0"}
        )

        returncode_off, stdout_off, stderr_off = self._run_mimic(param_file_off)
        self.assertEqual(returncode_off, 0, "Should run with reionization OFF")

        # Both should succeed (physics differences tested later)
        self.assertIn("ReionizationOn = 1", stdout_on, "ON mode logged")
        self.assertIn("ReionizationOn = 0", stdout_off, "OFF mode logged")

    def test_memory_safety(self):
        """Test that sage_infall doesn't leak memory."""
        print("\n  Testing memory safety...")

        param_file = self._create_test_param_file(
            output_name="sage_infall_memory",
            enabled_modules=["sage_infall"]
        )

        returncode, stdout, stderr = self._run_mimic(param_file)
        self.assertEqual(returncode, 0, "Execution should succeed")

        # Check for memory leak messages
        self.assertNotIn("Memory leak detected", stdout,
                        "Should not have memory leaks")
        self.assertNotIn("Memory leak detected", stderr,
                        "Should not have memory leaks in stderr")

    def test_execution_completes(self):
        """Test that full pipeline execution completes without errors."""
        print("\n  Testing full pipeline completion...")

        param_file = self._create_test_param_file(
            output_name="sage_infall_complete",
            enabled_modules=["sage_infall"],
            first_file=0,
            last_file=0  # Process single file
        )

        returncode, stdout, stderr = self._run_mimic(param_file)
        self.assertEqual(returncode, 0, "Pipeline should complete successfully")

        # Check for success messages
        self.assertIn("SAGE infall module initialized", stdout,
                     "Module initialization message")
        self.assertIn("SAGE infall module cleaned up", stdout,
                     "Module cleanup message")

    def test_multiple_module_pipeline(self):
        """Test that sage_infall works with other modules in pipeline."""
        print("\n  Testing multi-module pipeline...")

        # Test sage_infall with simple_cooling and simple_sfr
        param_file = self._create_test_param_file(
            output_name="sage_infall_multi",
            enabled_modules=["sage_infall", "simple_cooling", "simple_sfr"],
            module_params={
                "SageInfall_BaryonFrac": "0.17",
                "SimpleCooling_BaryonFraction": "0.15",
                "SimpleSFR_Efficiency": "0.02"
            }
        )

        returncode, stdout, stderr = self._run_mimic(param_file)
        self.assertEqual(returncode, 0,
                        "Multi-module pipeline should execute successfully")

        # Verify all modules initialized
        self.assertIn("SAGE infall module initialized", stdout)
        self.assertIn("Simple cooling module initialized", stdout)
        self.assertIn("Simple star formation rate module initialized", stdout)


if __name__ == '__main__':
    # Run tests
    unittest.main(verbosity=2)

File: src/modules/sage_infall/test_sage_infall.c
/**
 * @file    test_sage_infall.c
 * @brief   Software quality unit tests for sage_infall module
 *
 * Validates: Module lifecycle, memory safety, parameter handling, error handling
 * Phase: Phase 4.2 (SAGE Physics Module Implementation)
 *
 * This test validates software engineering aspects of the sage_infall module:
 * - Module registration and initialization
 * - Parameter reading and validation
 * - Memory allocation and cleanup (no leaks)
 * - Null pointer safety
 * - Property access patterns
 *
 * Test cases:
 *   - test_module_registration: Module registers correctly
 *   - test_module_initialization: Module init/cleanup lifecycle
 *   - test_parameter_reading: Module parameters read from config
 *   - test_memory_safety: No memory leaks during operation
 *   - test_property_access: Galaxy property access works correctly
 *
 * NOTE: Physics validation (reionization, infall calculations) deferred to Phase 4.3+
 *       when downstream modules (cooling, star formation) are implemented.
 *
 * @author  Mimic Development Team
 * @date    2025-11-12
 */

#include "framework/test_framework.h"
#include "../../core/module_registry.h"
#include "../../core/module_interface.h"
#include "sage_infall.h"
#include "../../include/types.h"
#include "../../include/proto.h"
#include "../../include/globals.h"
#include "../../util/error.h"
#include "../../util/memory.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

/* Test statistics (required for TEST_RUN macro) */
static int passed = 0;
static int failed = 0;

/* Track whether modules have been registered */
static int modules_registered = 0;

/* Test fixture: reset configuration state */
static void reset_config(void)
{
    memset(&MimicConfig, 0, sizeof(MimicConfig));
}

/* Test fixture: ensure modules are registered (only once) */
static void ensure_modules_registered(void)
{
    if (!modules_registered) {
        register_all_modules();
        modules_registered = 1;
    }
}

/* Test fixture: Set sage_infall parameters to defaults */
static void set_default_sage_infall_params(void)
{
    strcpy(MimicConfig.ModuleParams[0].module_name, "SageInfall");
    strcpy(MimicConfig.ModuleParams[0].param_name, "BaryonFrac");
    strcpy(MimicConfig.ModuleParams[0].value, "0.17");

    strcpy(MimicConfig.ModuleParams[1].module_name, "SageInfall");
    strcpy(MimicConfig.ModuleParams[1].param_name, "ReionizationOn");
    strcpy(MimicConfig.ModuleParams[1].value, "1");

    strcpy(MimicConfig.ModuleParams[2].module_name, "SageInfall");
    strcpy(MimicConfig.ModuleParams[2].param_name, "Reionization_z0");
    strcpy(MimicConfig.ModuleParams[2].value, "8.0");

    strcpy(MimicConfig.ModuleParams[3].module_name, "SageInfall");
    strcpy(MimicConfig.ModuleParams[3].param_name, "Reionization_zr");
    strcpy(MimicConfig.ModuleParams[3].value, "7.0");

    strcpy(MimicConfig.ModuleParams[4].module_name, "SageInfall");
    strcpy(MimicConfig.ModuleParams[4].param_name, "StrippingSteps");
    strcpy(MimicConfig.ModuleParams[4].value, "10");

    MimicConfig.NumModuleParams = 5;
}

/**
 * @test    test_module_registration
 * @brief   Test that sage_infall module registers correctly
 *
 * Expected: Module registration succeeds without errors
 * Validates: sage_infall_register() works, module appears in registry
 */
int test_module_registration(void)
{
    /* ===== SETUP ===== */
    reset_config();

    /* ===== EXECUTE ===== */
    ensure_modules_registered();

    /* ===== VALIDATE ===== */
    /* If we got here without crashing, registration succeeded */
    /* Module registry is internal, but we can test that module init works */

    return TEST_PASS;
}

/**
 * @test    test_module_initialization
 * @brief   Test module initialization and cleanup lifecycle
 *
 * Expected: Module init and cleanup succeed without errors or leaks
 * Validates: Module lifecycle management
 */
int test_module_initialization(void)
{
    /* ===== SETUP ===== */
    reset_config();
    init_memory_system(0);
    ensure_modules_registered();

    /* Set up minimal cosmology configuration */
    MimicConfig.Omega = 0.25;
    MimicConfig.OmegaLambda = 0.75;
    MimicConfig.Hubble_h = 0.73;

    /* Configure sage_infall module */
    strcpy(MimicConfig.EnabledModules[0], "sage_infall");
    MimicConfig.NumEnabledModules = 1;
    set_default_sage_infall_params();

    /* ===== EXECUTE ===== */
    int result = module_system_init();

    /* ===== VALIDATE ===== */
    TEST_ASSERT(result == 0, "Module system initialization should succeed");

    /* ===== CLEANUP ===== */
    module_system_cleanup();
    check_memory_leaks();

    return TEST_PASS;
}

/**
 * @test    test_parameter_reading
 * @brief   Test that module parameters are read correctly from configuration
 *
 * Expected: Module reads all 5 parameters successfully
 * Validates: Parameter reading infrastructure works
 */
int test_parameter_reading(void)
{
    /* ===== SETUP ===== */
    reset_config();
    init_memory_system(0);
    ensure_modules_registered();

    /* Set up configuration */
    MimicConfig.Omega = 0.25;
    MimicConfig.OmegaLambda = 0.75;
    MimicConfig.Hubble_h = 0.73;

    /* Configure with non-default values */
    strcpy(MimicConfig.EnabledModules[0], "sage_infall");
    MimicConfig.NumEnabledModules = 1;

    strcpy(MimicConfig.ModuleParams[0].module_name, "SageInfall");
    strcpy(MimicConfig.ModuleParams[0].param_name, "BaryonFrac");
    strcpy(MimicConfig.ModuleParams[0].value, "0.20");

    strcpy(MimicConfig.ModuleParams[1].module_name, "SageInfall");
    strcpy(MimicConfig.ModuleParams[1].param_name, "ReionizationOn");
    strcpy(MimicConfig.ModuleParams[1].value, "0");

    MimicConfig.NumModuleParams = 2;

    /* ===== EXECUTE ===== */
    int result = module_system_init();

    /* ===== VALIDATE ===== */
    TEST_ASSERT(result == 0, "Module should initialize with custom parameters");
    /* If init succeeded, parameters were read and validated */

    /* ===== CLEANUP ===== */
    module_system_cleanup();
    check_memory_leaks();

    return TEST_PASS;
}

/**
 * @test    test_memory_safety
 * @brief   Test that module doesn't leak memory during normal operation
 *
 * Expected: No memory leaks after init, process, cleanup cycle
 * Validates: Memory management in module
 */
int test_memory_safety(void)
{
    /* ===== SETUP ===== */
    reset_config();
    init_memory_system(0);
    ensure_modules_registered();

    MimicConfig.Omega = 0.25;
    MimicConfig.OmegaLambda = 0.75;
    MimicConfig.Hubble_h = 0.73;

    strcpy(MimicConfig.EnabledModules[0], "sage_infall");
    MimicConfig.NumEnabledModules = 1;
    set_default_sage_infall_params();

    /* ===== EXECUTE ===== */
    int result = module_system_init();
    TEST_ASSERT(result == 0, "Module initialization should succeed");

    /* ===== VALIDATE ===== */
    /* Module initialized successfully without memory leaks */
    /* (Full pipeline processing tested in integration tests) */

    /* ===== CLEANUP ===== */
    module_system_cleanup();
    check_memory_leaks();

    return TEST_PASS;
}

/**
 * @test    test_property_access
 * @brief   Test that module can safely access galaxy properties
 *
 * Expected: Property access doesn't crash, handles zero/null gracefully
 * Validates: Property access patterns in module
 */
int test_property_access(void)
{
    /* ===== SETUP ===== */
    init_memory_system(0);

    /* Create test halo and galaxy with various property states */
    struct Halo test_halo;
    memset(&test_halo, 0, sizeof(test_halo));

    struct GalaxyData test_galaxy;
    memset(&test_galaxy, 0, sizeof(test_galaxy));

    /* Set some realistic values */
    test_halo.Mvir = 100.0;  /* 10^12 Msun/h */
    test_halo.Type = 0;  /* Central */
    test_halo.SnapNum = 63;
    test_halo.galaxy = &test_galaxy;

    test_galaxy.StellarMass = 10.0;
    test_galaxy.ColdGas = 5.0;
    test_galaxy.HotGas = 15.0;
    test_galaxy.EjectedMass = 2.0;
    test_galaxy.ICS = 0.5;
    test_galaxy.MetalsHotGas = 0.3;
    test_galaxy.MetalsEjectedMass = 0.04;
    test_galaxy.MetalsICS = 0.01;

    /* ===== VALIDATE ===== */
    /* Test that halo properties can be accessed without crashing */
    TEST_ASSERT(test_halo.Mvir > 0.0, "Mvir should be accessible");
    TEST_ASSERT(test_halo.Type == 0, "Type should be accessible");
    TEST_ASSERT(test_halo.galaxy != NULL, "Galaxy pointer should be accessible");

    /* Test that galaxy properties can be accessed */
    TEST_ASSERT(test_galaxy.StellarMass >= 0.0, "StellarMass should be non-negative");
    TEST_ASSERT(test_galaxy.HotGas >= 0.0, "HotGas should be non-negative");

    /* Test with zero values (edge case) */
    struct GalaxyData zero_galaxy;
    memset(&zero_galaxy, 0, sizeof(zero_galaxy));
    TEST_ASSERT(zero_galaxy.HotGas == 0.0, "Zero-initialized galaxy should have HotGas=0");
    TEST_ASSERT(zero_galaxy.ColdGas == 0.0, "Zero-initialized galaxy should have ColdGas=0");

    /* ===== CLEANUP ===== */
    check_memory_leaks();

    return TEST_PASS;
}

/**
 * @brief   Main test runner
 *
 * Executes all sage_infall software quality tests and reports results
 */
int main(void)
{
    printf("\n");
    printf("========================================\n");
    printf("  sage_infall Software Quality Tests\n");
    printf("========================================\n");
    printf("\n");
    printf("NOTE: Physics validation deferred to Phase 4.3+\n");
    printf("      These tests validate software engineering aspects only.\n");
    printf("\n");

    /* Run all tests */
    TEST_RUN(test_module_registration);
    TEST_RUN(test_module_initialization);
    TEST_RUN(test_parameter_reading);
    TEST_RUN(test_memory_safety);
    TEST_RUN(test_property_access);

    /* Print summary and return result */
    TEST_SUMMARY();
    return TEST_RESULT();
}
